### PR TITLE
Properly install pod targets with test specs within subspecs

### PR DESCRIPTION
**Pull request by dnkoutso**

### CODE REVIEW

**Comment by CocoaPodsBarista** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please include a CHANGELOG entry to credit yourself! <br />
You can find it at <a href="https://github.com/CocoaPods/CocoaPods/blob/master/CHANGELOG.md">CHANGELOG.md</a>.</td>
    </tr>
  </tbody>
</table>

Here's an example of your CHANGELOG entry:

```markdown
* Properly install pod targets that are part of subspecs  
  [dnkoutso](https://github.com/dnkoutso)
  [#issue_number](https://github.com/CocoaPods/CocoaPods/issues/issue_number)
```

*note*: There are two invisible spaces after the entry's text.


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: lib/cocoapods/installer/analyzer.rb [hunk 1]
```diff
@@ -429,17 +429,28 @@ def generate_target(target_definition, pod_targets)
...
         end
 
-        target.pod_targets = pod_targets.select do |pod_target|
-          target_definition_included = pod_target.target_definitions.include?(target_definition)
-          explicitly_defined_in_target_definition = target_definition.non_inherited_dependencies.map(&:name).include?(pod_target.name)
-          next false unless target_definition_included
-          next true if explicitly_defined_in_target_definition
-          !pod_target_test_only?(pod_target, pod_targets)
-        end
+        target.pod_targets = filter_pod_targets_for_target_definition(pod_targets, target_definition)
```
Comment (dnkoutso) [inline lib/cocoapods/installer/analyzer.rb]: extracted into a method so I can test it easier.

File: lib/cocoapods/installer/analyzer.rb [hunk 2]
```diff
@@ -457,8 +468,12 @@ def pod_target_test_only?(pod_target, pod_targets)
         if @test_pod_target_analyzer_cache.key?(key)
           return @test_pod_target_analyzer_cache[key]
         end
-        source = pod_targets.any? { |pt| pt.dependent_targets.map(&:name).include?(name) }
-        test = pod_targets.any? { |pt| pt.test_dependent_targets.map(&:name).include?(name) }
+        source = pod_targets.any? do |pt|
+          pt.dependent_targets.map(&:name).include?(name)
+        end
+        test = pod_targets.any? do |pt|
+          pt.test_dependent_targets.reject { |dpt| dpt.name == pt.name }.map(&:name).include?(name)
```
Comment (dnkoutso) [inline lib/cocoapods/installer/analyzer.rb]: We intentionally ignore test dependencies to ourselves. This was the bug. A test spec technically always depends on the parent pod target so when we added a dependency to an aggregate target that has 0 other dependencies it was mistakenly considered as test only.

Test case below to cover for this.

File: lib/cocoapods/installer/analyzer.rb [hunk 3]
```diff
@@ -429,17 +429,28 @@ def generate_target(target_definition, pod_targets)
...
+      #        the array of pod targets to check against
+      #
+      # @param [TargetDefinition] target_definition
+      #        the target definition to use as the base for filtering
+      #
+      # @return [Array<PodTarget>] the filtered list of pod targets.
+      #
+      def filter_pod_targets_for_target_definition(pod_targets, target_definition)
+        pod_targets.select do |pod_target|
+          pod_target.target_definitions.include?(target_definition) && !pod_target_test_only?(pod_target, pod_targets)
```
Comment (dnkoutso) [inline lib/cocoapods/installer/analyzer.rb]: a pod target should be part of an aggregate target if:

1) It's included in its target definition (as it has always been before test specs)
2) It is not used only by tests.

For example, 'OCMock' IS part of the target definition BUT it is only referenced by only pod targets as their test dependency therefore it should NOT be part of the aggregate target.

**Comment by dnkoutso** [comment]
@DanToml with this you can release 1.3.0

File: lib/cocoapods/installer/analyzer.rb (hunk 3 above)
Comment (DanToml) [inline lib/cocoapods/installer/analyzer.rb]: nice :+1:

File: lib/cocoapods/installer/analyzer.rb (hunk 2 above)
Comment (DanToml) [inline lib/cocoapods/installer/analyzer.rb]: good catch :+1:

**Pull request merged**